### PR TITLE
[enh] Summary of the steps to contribute/dev on the core of Yunohost

### PR DESCRIPTION
**Pull request by alexAubin**

### CODE REVIEW

**Comment by M5oul** [comment]
Thanks for this awesome work.
Could you add links toward this page?
At least for [en and fr sitemap pages](https://yunohost.org/#/sitemap_fr).

**Comment by M5oul** [comment]
Someone from @YunoHost/devs and @YunoHost/doc teams need to review this work.

File: dev.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  /var/cache/moulinette/actionsmap/yunohost.pkl`
+
+## Adding stuff to the Yunohost web administration interface
+
+- Work in `/vagrant/yunohost-admin/src/`
+
+- Make sure the web admin interface of your development environment works, by
+  opening `https://domain.tld/yunohost/admin`. If not, run `npm install` and
+  `npm run build` in `yunohost-admin/src/`. [Not sure about this, to be checked]
+
```
Comment (zamentur) [inline dev.md]: You juste need to use:
./ynh-dev use-git yunohost-admin

This command run gulp, so each time you modify source js file, it recompiles the code and you can use it by refreshing (CTRL + F5) your web administration.
To stop the command just do CTRL + C

File: dev.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#### Don't forget
+
+- Each time you edit the actionsmap, you should restart the yunohost-api :
+  ```service yunohost-api restart```
+  (You'll need to retype your admin password in the web interface)
+
+- Each time you edit something in the `js` folder, you should re-build with npm :
+  ```npm run build-dev``` 
+  and clear the cache of your browser.
+
```
Comment (zamentur) [inline dev.md]: See my comment L65

File: dev.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+- Each time you edit something in the `views` folder, you should clear the cache
+  of your browser
+
+
+## Your feature is working and you want it to be integrated in Yunohost
+
+- Fork the relevant repo on Github, and commit stuff to a new branch with a name
+  like `enh-name-of-feature`.
+
```
Comment (zamentur) [inline dev.md]: I suggest "enh-IDREDMINETICKET-name-of-feature" or "fix-IDREDMINETICKET-what-goes-wrong"

**Comment by M5oul** [comment]
Seems good to me from doc and dev points of views.
Great work!
It will be very usefull for newcomers 👍

**Comment by alexAubin** [comment]
Implement remarks from ljf. I'm still not entirely sure about some points though (especially the workflow for yunohost-admin), so feel free to rewrite some parts !

File: dev.md (hunk 1 above)
Comment (M5oul) [inline dev.md]: I tried https://github.com/YunoHost/yunohost-admin/pull/136
And get this issue:
```bash
y-uniq@1.0.3, replace-ext@0.0.1, has-gulplog@0.1.0, fancy-log@1.2.0, gulplog@1.0.0, lodash.template@3.6.2, vinyl@0.5.3, through2@2.0.1, multipipe@0.1.2, dateformat@1.0.12)
└── vinyl-fs@0.3.14 (strip-bom@1.0.0, graceful-fs@3.0.11, defaults@1.0.3, vinyl@0.4.6, mkdirp@0.5.1, through2@0.6.5, glob-stream@3.1.18, glob-watcher@0.0.6)
No passwd entry for user 'vagrant'
No passwd entry for user 'vagrant'
Using Git repository for yunohost-admin
No passwd entry for user 'vagrant'
```

With the admin broken:

Comment (M5oul) [inline dev.md]: Finally, [found a solution](https://dev.yunohost.org/issues/641#note-1) for non usage of vagrant box.

```bash
If not, run `npm install` and
+  `npm run build` in `yunohost-admin/src/`. [Not sure about this, to be checked]
```
Could be removed.

File: dev.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
+## Contributing to the Yunohost core
+
+You wish to add a new feature in the Yunohost core, but don't know how to
+proceed ? This guide takes you through the various steps of the development and
```
Comment (theodorefaure) [inline dev.md]: how to proceed? (without space)

File: dev.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+- Work in `/vagrant/yunohost-admin/src/`.
+
+- Run `/vagrant/ynh-dev use-git yunohost-admin`. It launches gulp, such as each 
+  time you modify sources, it recompiles the code and you can use it by 
+  refreshing (Ctrl+F5) your web administration. To stop the command, just do Ctrl+C.
+
+- The web interface uses the API to interact with Yunohost. The API
+  commands/requests are also defined via the actionsmap. For instance, accessing
+  the page ```https://domain.tld/yunohost/api/users``` corresponds to a `GET
+  /domains` requests on the Yunohost API. It is mapped to the function
```
Comment (theodorefaure) [inline dev.md]: shouldn't it be "GET /users" instead?

File: dev_fr.md [hunk 6]
```diff
@@ -1,91 +1,164 @@
...
-172.17.0.1   exemple.local
-```
-
-##### Déployer les sources au fur et à mesure des modifications
-```bash
-root@yunohost:/yunohost/# ./ynh-dev watch
-```
-
-Astuce : dans le cas de modification sur yunohost-admin, cette commande est très pratique couplée avec un `gulp watch` sur la machine hôte.
+## Contribuer au coeur de Yunohost
```
Comment (theodorefaure) [inline dev_fr.md]: c**œ**ur de Yuno**H**ost

File: dev_fr.md [hunk 7]
```diff
@@ -1,91 +1,164 @@
...
+
+Vous souhaitez ajouter une nouvelle fonctionnalité au coeur de Yunohost, mais ne
```
Comment (theodorefaure) [inline dev_fr.md]: c**œ**ur de Yuno**H**ost

File: dev_fr.md [hunk 8]
```diff
@@ -1,91 +1,164 @@
...
+**Venez dire coucou sur le [salon de
+dev](xmpp:dev@conference.yunohost.org?join)** ! Si vous n'avez pas de client
+XMPP, vous devriez pouvoir vous connecter à l'aide du widget en bas de la page.
+
+### Mettre en place un environnement de développement
+
+- **Utilisez [ynh-dev](https://github.com/YunoHost/ynh-dev)** (voir le README)
+  pour mettre en place un environnement de développement - en local sur une
+  machine virtuel, ou bien sur un VPS.
+  Ceci installera une instance fonctionelle de Yunohost, en utilisant
```
Comment (theodorefaure) [inline dev_fr.md]: Yuno**H**ost

File: dev_fr.md [hunk 9]
```diff
@@ -1,91 +1,164 @@
...
+  temps réel, et de commiter et push/pull directement depuis cet environnement.
+
+- **Implémentez et testez votre fonctionnalité**. Suivant ce sur quoi vous
+  voulez travailler :
+   - **Coeur Python/ligne de comande** : allez dans `/vagrant/yunohost/`
+   - **Interface d'administration web** : allez dans `/vagrant/yunohost-admin/`
+   - Vous pouvez aussi travailler sur les autres projets liés sur lesquels
+     s'appuie Yunohost (SSOwat, moulinette) de facon similaire.
+
+### Travailler sur le coeur Python / ligne de commande
```
Comment (theodorefaure) [inline dev_fr.md]: c**œ**ur

File: dev_fr.md [hunk 10]
```diff
@@ -1,91 +1,164 @@
...
+
+- Allez dans `/vagrant/yunohost/`.
+
+- Executez `/vagrant/ynh-dev use-git yunohost`.
+
+- Le fichier actionsmap (`data/actionsmap/yunohost.yml`) défini les différentes
```
Comment (theodorefaure) [inline dev_fr.md]: défini**t**

File: dev_fr.md [hunk 11]
```diff
@@ -1,91 +1,164 @@
...
+  catégories, actions et arguments de la ligne de commande Yunohost. Choisissez
+  comment vous voulez que les utilisateurs utilise votre fonctionnalité, et
```
Comment (theodorefaure) [inline dev_fr.md]: utilise**nt**

File: dev_fr.md [hunk 12]
```diff
@@ -1,91 +1,164 @@
...
+
+- Yunohost essaye de suivre le style de code [pep8](http://pep8.org/). Des
+  outils existent pour vérifier automatiquement la conformité du code.
+
+- Mettre un `_` devant les noms des fonctions "privées".
+
+##### N'oubliez pas
+
+- (Peut-être plus nécessaire) À chaque fois que vous modifiez l'actionsmap, il
+  faut forcer le rafraichissement du cache avec :
```
Comment (theodorefaure) [inline dev_fr.md]: rafra**î**chissement

File: dev_fr.md [hunk 13]
```diff
@@ -1,91 +1,164 @@
...
+  `rm /var/cache/moulinette/actionsmap/yunohost.pkl`
+
+### Travailler sur l'interface d'administration web
+
+- Allez dans `/vagrant/yunohost-admin/src/`.
+
+- Executez `/vagrant/ynh-dev use-git yunohost-admin`. Ceci lance gulp, de sorte
```
Comment (theodorefaure) [inline dev_fr.md]: Ex**é**cutez

File: dev_fr.md [hunk 14]
```diff
@@ -1,91 +1,164 @@
...
+  à ce que à chaque fois que vous modifiez les sources, il recompilera le code
```
Comment (theodorefaure) [inline dev_fr.md]: "**de sorte qu'à** chaque fois que"

File: dev_fr.md [hunk 15]
```diff
@@ -1,91 +1,164 @@
...
+  (js) et vous pourrez voir les changements dans le navigateur web (Ctrl+F5).
+  Pour stopper la commande, faites simplement Ctrl+C.
+
+- L'interface web utilise une API pour communiquer avec Yunohost. Les
```
Comment (theodorefaure) [inline dev_fr.md]: Yuno**H**ost

File: dev_fr.md [hunk 16]
```diff
@@ -1,91 +1,164 @@
...
+  commandes/requêtes de l'API sont également définie dans l'actionsmap. Par
```
Comment (theodorefaure) [inline dev_fr.md]: définie**s**

File: dev_fr.md [hunk 17]
```diff
@@ -1,91 +1,164 @@
...
+  exemple, accéder à la page ```https://domain.tld/yunohost/api/users```
+  corresponds à une requete `GET /domains` vers l'API yunohost. Cette requete
```
Comment (theodorefaure) [inline dev_fr.md]: correspond

Comment (theodorefaure) [inline dev_fr.md]: requ**ê**te `GET /users` (non ?)

File: dev_fr.md [hunk 18]
```diff
@@ -1,91 +1,164 @@
...
+  est mappée sur `user_list()`. Accéder à cette URL devrait afficher le json
+  retourné par cette fonction. Les requetes `GET' sont typiquement destinés à
```
Comment (theodorefaure) [inline dev_fr.md]: requ**ê**tes

Comment (theodorefaure) [inline dev_fr.md]: destiné**e**s

File: dev_fr.md [hunk 19]
```diff
@@ -1,91 +1,164 @@
...
+  demander de l'information au serveur, tandis que les requetes `POST` sont
```
Comment (theodorefaure) [inline dev_fr.md]: requ**ê**tes

File: dev_fr.md [hunk 20]
```diff
@@ -1,91 +1,164 @@
...
+  destinées à demander au serveur de modifier/changer des informations ou de
+  réaliser des actions.
+
+- `js/yunohost/controllers` contiens les parties javascript, et défini quelles
```
Comment (theodorefaure) [inline dev_fr.md]: contien**t**

Comment (theodorefaure) [inline dev_fr.md]: défini**t**

File: dev_fr.md [hunk 21]
```diff
@@ -1,91 +1,164 @@
...
+  locales autres que en.json, la traduction sera faite avec 
+  [weblate](https://translate.yunohost.org/) !
+
+##### N'oubliez pas
+
+- À chaque modification de l'actionsmap, il faut redémarrer l'API yunohost :
+  ```service yunohost-api restart```
+  (Il faudra retaper le mot de passe administrateur dans l'interface web)
+
+- Il faudra peut-être régulièrement forcer le rafraichissement du cache
```
Comment (theodorefaure) [inline dev_fr.md]: rafra**î**chissement

File: dev_fr.md [hunk 22]
```diff
@@ -1,91 +1,164 @@
...
+  navigateur pour propager correctement le javascript et/ou HTML (à chaque fois
+  que l'on change quelque chose dans `js` ou `views`, donc).
+
+
+### Votre fonctionnalité est prête et vous souhaitez qu'elle soit intégrée dans Yunohost 
+
+- Forkez le dépòt correspondant sur Github, et commitez votre changements dans
```
Comment (theodorefaure) [inline dev_fr.md]: votre **changement** (car singulier)

File: dev_fr.md [hunk 23]
```diff
@@ -1,91 +1,164 @@
...
+dev](xmpp:dev@conference.yunohost.org?join)** ! Si vous n'avez pas de client
+XMPP, vous devriez pouvoir vous connecter à l'aide du widget en bas de la page.
+
+### Mettre en place un environnement de développement
+
+- **Utilisez [ynh-dev](https://github.com/YunoHost/ynh-dev)** (voir le README)
+  pour mettre en place un environnement de développement - en local sur une
+  machine virtuel, ou bien sur un VPS.
+  Ceci installera une instance fonctionelle de Yunohost, en utilisant
+  directement les dépôts git à l'aide de liens symboliques. De cette facon, il
```
Comment (theodorefaure) [inline dev_fr.md]: fa**ç**on

File: dev_fr.md [hunk 24]
```diff
@@ -1,91 +1,164 @@
...
+  vous sera possible de modifier les fichiers, de tester les changements en
+  temps réel, et de commiter et push/pull directement depuis cet environnement.
+
+- **Implémentez et testez votre fonctionnalité**. Suivant ce sur quoi vous
+  voulez travailler :
+   - **Coeur Python/ligne de comande** : allez dans `/vagrant/yunohost/`
```
Comment (theodorefaure) [inline dev_fr.md]: C**œ**ur

File: dev_fr.md [hunk 25]
```diff
@@ -1,91 +1,164 @@
...
+- À chaque modification de l'actionsmap, il faut redémarrer l'API yunohost :
+  ```service yunohost-api restart```
+  (Il faudra retaper le mot de passe administrateur dans l'interface web)
+
+- Il faudra peut-être régulièrement forcer le rafraichissement du cache
+  navigateur pour propager correctement le javascript et/ou HTML (à chaque fois
+  que l'on change quelque chose dans `js` ou `views`, donc).
+
+
+### Votre fonctionnalité est prête et vous souhaitez qu'elle soit intégrée dans Yunohost
```
Comment (theodorefaure) [inline dev_fr.md]: Yuno**H**ost

File: sitemap.md [hunk 26]
```diff
@@ -80,16 +80,18 @@
...
+                   * [Manifest](packaging_apps_manifest_en)
+                   * [Scripts](packaging_apps_scripts_en)
+                   * [Arguments management](packaging_apps_arguments_management_en)
+                   * [Nginx configuration](packaging_apps_nginx_conf_en)
+                   * [Multi-instance](packaging_apps_multiinstance_en)
+                   * [Helpers](packaging_apps_helpers_en)
+                   * [Trap usage](packaging_apps_trap_fr)
+               * [Applications in progress](/apps_in_progress_en)
+               * [Wishlist of applications](/apps_wishlist_en)
+           * [Contributing to the Yunohost core](/dev)
```
Comment (theodorefaure) [inline sitemap.md]: Yuno**H**ost

File: dev.md [hunk 27]
```diff
@@ -0,0 +1,155 @@
+## Contributing to the Yunohost core
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+
+You wish to add a new feature in the Yunohost core, but don't know how to
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+If you're looking for stuff to implement or fix, the bug-tracker is 
+[here](https://dev.yunohost.org/issues/) !
+
+**Come say hi to us in the [dev chatroom](xmpp:dev@conference.yunohost.org?join)** ! If you don't have an XMPP client, you can join using the widget at the bottom right of this page.
+
+### Setting up a development environment
+
+- **Use [ynh-dev](https://github.com/YunoHost/ynh-dev)** (see the README) to
+  setup a development environment - locally in a virtual machine, or on a VPS.
+  This will setup a working Yunohost instance, using directly the git repositories
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+  (with symlinks). That way, you will be able to edit files, test your changes in real
+  time, commit stuff and push/pull directly from your development environment.
+
+- **Implement and test your feature**. Depending on what you want to develop, you
+  will want to :
+   - **Python/CLI core** : work in `/vagrant/yunohost/`
+   - **Web administration interface** : work in `/vagrant/yunohost-admin/`
+   - You can also work on the other projects on which Yunohost is built
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+  time, commit stuff and push/pull directly from your development environment.
+
+- **Implement and test your feature**. Depending on what you want to develop, you
+  will want to :
+   - **Python/CLI core** : work in `/vagrant/yunohost/`
+   - **Web administration interface** : work in `/vagrant/yunohost-admin/`
+   - You can also work on the other projects on which Yunohost is built 
+     (SSOwat, moulinette) in similar ways
+
+### Working on the Yunohost Python/CLI core
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+##### Helpers / coding style
+
+- To handle exceptions, you should raise some `MoulinetteError()`
+
+- To help with internationalizing the messages, use `m18n.n('some-message-id')`
+  and put your string in `locales/en.json`. You can also put arguments and use
+  them in the string with `{{some-argument:s}}`. Don't edit other locales files,
+  this will be done using [weblate](https://translate.yunohost.org/) !
+
+- Yunohost tries to follow the [pep8](http://pep8.org/) coding style. Tools
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 33]
```diff
@@ -0,0 +1,155 @@
...
+
+- Name of "private" functions should start with a `_`
+
+##### Don't forget
+
+- (Might not be necessary anymore) Each time you edit the actionsmap, you should
+  force the refresh of the cache with `rm
+  /var/cache/moulinette/actionsmap/yunohost.pkl`
+
+### Working on the Yunohost web administration interface
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 34]
```diff
@@ -0,0 +1,155 @@
...
+
+- Work in `/vagrant/yunohost-admin/src/`.
+
+- Run `/vagrant/ynh-dev use-git yunohost-admin`. It launches gulp, such as each 
+  time you modify sources, it recompiles the code and you can use it by 
+  refreshing (Ctrl+F5) your web administration. To stop the command, just do Ctrl+C.
+
+- The web interface uses the API to interact with Yunohost. The API
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md (hunk 5 above)
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev.md [hunk 35]
```diff
@@ -0,0 +1,155 @@
...
+
+- Each time you edit the actionsmap, you should restart the yunohost-api :
+  ```service yunohost-api restart```
+  (You'll need to retype your admin password in the web interface)
+
+- You might need to force-clear the cache of your browser sometimes to refresh 
+  the javascript and/or html (so each time you edit something in `js` or `views`).
+
+
+### Your feature is ready and you want it to be integrated in Yunohost
```
Comment (theodorefaure) [inline dev.md]: Yuno**H**ost

File: dev_fr.md [hunk 36]
```diff
@@ -1,91 +1,164 @@
...
+  directement les dépôts git à l'aide de liens symboliques. De cette facon, il
+  vous sera possible de modifier les fichiers, de tester les changements en
+  temps réel, et de commiter et push/pull directement depuis cet environnement.
+
+- **Implémentez et testez votre fonctionnalité**. Suivant ce sur quoi vous
+  voulez travailler :
+   - **Coeur Python/ligne de comande** : allez dans `/vagrant/yunohost/`
+   - **Interface d'administration web** : allez dans `/vagrant/yunohost-admin/`
+   - Vous pouvez aussi travailler sur les autres projets liés sur lesquels
+     s'appuie Yunohost (SSOwat, moulinette) de facon similaire.
```
Comment (theodorefaure) [inline dev_fr.md]: Yuno**H**ost

File: dev_fr.md [hunk 37]
```diff
@@ -1,91 +1,164 @@
...
+
+### Travailler sur le coeur Python / ligne de commande
+
+- Allez dans `/vagrant/yunohost/`.
+
+- Executez `/vagrant/ynh-dev use-git yunohost`.
+
+- Le fichier actionsmap (`data/actionsmap/yunohost.yml`) défini les différentes
+  catégories, actions et arguments de la ligne de commande Yunohost. Choisissez
```
Comment (theodorefaure) [inline dev_fr.md]: Yuno**H**ost

File: dev_fr.md [hunk 38]
```diff
@@ -1,91 +1,164 @@
...
+
+- Pour gérer les exceptions, il existe un type `MoulinetteError()`
+
+- Pour aider avec l'internationalisation des messages, utilisez `m18n.n('some-message-id')`
+  et mettez le message correspondant dans `locales/en.json`. Vous pouvez aussi
+  utiliser des arguments pour construire les messages, avec `{{some-argument:s}}`. 
+  Ne modifiez pas de fichiers de locales autres que en.json, la traduction sera
+  faite avec [weblate](https://translate.yunohost.org/) !
+
+- Yunohost essaye de suivre le style de code [pep8](http://pep8.org/). Des
```
Comment (theodorefaure) [inline dev_fr.md]: Yuno**H**ost

File: sitemap_fr.md [hunk 39]
```diff
@@ -103,6 +103,7 @@
                  * [Usage de trap](packaging_apps_trap_fr)
               * [Packages d’applications en cours de développement](/apps_in_progress_fr)
               * [Liste de proposition d’applications](/apps_wishlist_fr)
+           * [Contribuer au coeur de Yunohost](/dev_fr)
```
Comment (theodorefaure) [inline sitemap_fr.md]: c**œ**ur de Yuno**H**ost

**Comment by theodorefaure** [comment]
Amazing work @alexAubin 👍 My review is mostly about typos. 
@YunoHost/devs could you check line 62 : `Might not be necessary anymore`

**Comment by alexAubin** [comment]
Thanks for the feedback 😄 !

**Pull request merged**